formats/format_aac.c

```diff
@@ -0,0 +1,611 @@
+		encoder_config->inputFormat = FAAC_INPUT_16BIT;
+		encoder_config->outputFormat = 1;
+		encoder_config->bitRate = sample_rate * 2;
+		encoder_config->bandWidth = 4000;
+		encoder_config->aacObjectType = LOW;
+		encoder_config->mpegVersion = MPEG4;
+		encoder_config->useTns = 1;
+		encoder_config->useLfe = 0;
+		encoder_config->jointmode = 1;
+		encoder_config->quantqual = 50;
+		encoder_config->pnslevel = 4;
+		faacEncSetConfiguration(p->encoder, encoder_config);
+
+		ast_debug(3, "Encoder expects %ld input samples per encode() call with %d channels, max output buffer size %ld bytes",
+			p->inputSamples,
+			CHANNELS,
+			p->maxOutputBytes
+		);
+	}
+	return p->encoder_counter;
+}
+
+static void aac_decoder_init(struct aac_private *p, unsigned long sample_rate)
+{
+	NeAACDecConfigurationPtr decoder_config;
+
+	p->decoder = NeAACDecOpen();
+	decoder_config = NeAACDecGetCurrentConfiguration(p->decoder);
+	/* Low-Complexity profile */
+	decoder_config->defObjectType = LC;
+	decoder_config->defSampleRate = sample_rate;
+	/* Only does 5.1 -> stereo downmixing, ... */
+	decoder_config->downMatrix = 1;
+	/* ... therefor we need this for proper downmixing instead of FAAD_FMT_16BIT */
+	decoder_config->outputFormat = FAAD_FMT_DOUBLE;	
+	decoder_config->dontUpSampleImplicitSBR = 1;
+	NeAACDecSetConfiguration(p->decoder, decoder_config);
+
+	return;
+}
+
+static void aac_decoder_deinit(struct ast_filestream *s)
+{
+	struct aac_private *p = s->_private;
+
+	if (p->decoder_counter != 0) {
+		NeAACDecClose(p->decoder);
+	}
+
+	return;
+}
+
+static int aac_encoder_deinit(struct ast_filestream *s)
+{
+	int encoderLastOutputBufferBytes = 0;
+	struct aac_private *p = s->_private;
+
+	if (p->encoder_counter != 0) {
+		if (p->edatalen > 0) {
+			aac_encoder_flush(s, p->esamples);
+		}
+		ast_free(p->ebuffer);
+		unsigned char encoderLastOutputBuffer[p->maxOutputBytes];
+reflush:
+		encoderLastOutputBufferBytes = faacEncEncode(p->encoder, NULL, 0, encoderLastOutputBuffer, p->maxOutputBytes);
+		if (encoderLastOutputBufferBytes > 0) {
+			fwrite(encoderLastOutputBuffer, 1, encoderLastOutputBufferBytes, s->f);
+			ast_debug(3, "Encoder last wrote: %d bytes", encoderLastOutputBufferBytes);
+			goto reflush;
+		}
+		faacEncClose(p->encoder);
+	}
+
+	return encoderLastOutputBufferBytes;
+}
+
+static int aac_open(struct ast_filestream *s)
+{
+	struct aac_private *p = s->_private;
+
+	/* decoder */
+	p->decoder_counter = 0;
+
+	/* encoder */
+	p->encoder_counter = 0;
+
+	return 0;
+}
+
+
+static void aac_close(struct ast_filestream *s)
+{
+	/* decoder */
+	aac_decoder_deinit(s);
+
+	/* encoder */
+	aac_encoder_deinit(s);
+
+	return;
+}
+
+static void aac_decode(struct aac_private *p, long offset)
+{
+	NeAACDecFrameInfo decoderInfo;
+
+	double * unsafe_buffer;
+	double sample;
+	long double sum;
+	unsigned int channels;
+
+	unsafe_buffer = NeAACDecDecode(p->decoder, &decoderInfo, p->dinput + offset, p->dconsumed);
```

Comment:
No, I've checked various implementations of this (eg.: [this](https://github.com/gypified/libfaad/blob/master/decode_example.c) or [this](https://github.com/Distrotech/faad2/blob/master/aacDECdrop/decode.c#L225) and even it's [docs](https://github.com/knik0/faad2/tree/master/docs).

Reply:
Also never told the size to that pointer to compiler. Which can not even be figured out. It would be -in this case- sizeof(double) * decoderInfo.channels * decoderInfo.samples. The latter two is only set during the call to NeAACDecDecode(), so it's not possible to pre-allocate memory.